services/state/token/account.proto

```diff
@@ -189,29 +189,64 @@ message Account {
      * and the time the system task actually auto-renews it.
      */
     bool expired_and_pending_removal = 31;
+    /**
+     * The first key in the doubly-linked list of this contract's storage mappings;
+     * It will be null if if the account is not a contract or the contract has no storage mappings.
+     */
+    Uint256Value first_contract_storage_key = 32;
 }
 
 /**
- * Allowance granted by this account to another account for a specific token.
+ * Allowance granted by this account to a spender for a specific non-fungible token
+ * using ApproveForAll. This allows spender to spend all serial numbers for the given
+ * non-fungible token number.
  */
-message AccountTokenAllowance {
+message AccountApprovalForAllAllowance {
     int64 token_num = 1;
-    int64 account_num = 2;
+    int64 spender_num = 2;
 }
 
 /**
  * Allowance granted by this account to another account for a specific fungible token.
  * This also contains the amount of the token that is approved for the account.
+ * This allows spender to spend the amount of tokens approved for the account.
  */
 message AccountFungibleTokenAllowance {
-    AccountTokenAllowance token_allowance_key = 1;
-    int64 amount = 2;
+    int64 token_num = 1;
+    int64 spender_num = 2;
+    int64 amount = 3;
 }
 
 /**
  * Allowance granted by this account to another account for an amount of hbars.
+ * This allows spender to spend the amount of hbars approved for the account.
  */
 message AccountCryptoAllowance {
-    int64 account_num = 1;
+    int64 spender_num = 1;
     int64 amount = 2;
 }
+/**
+ * Message representing a uint256 value.
+ */
+message Uint256Value {
+    /**
+     * The number low-order bytes in the 256-bits that contain ones
+     */
+    uint32 non_zero_bits = 1;
```

Comment:
Have we adopted a policy on use of the `uint` protobuf variants? (Since in Java we always end up with signed integers, I think at some point Leemon suggested we use strictly `int32` and `int64`.)

Reply:
Since this message says `Uint256Value` I have used `uint` protobuf types. Will change to int32. Thanks @tinker-michaelj

Reply:
The non_zero_bits value could be signed, as it's not part of the 256-bit value.
If we used int64 for the segments, however, the bits get shifted around for 2s compliment when the high bit is set, so we can't easily concatenate the bits into a single 256 bit value without some tricky manipulation.
We could use a single bytes field and try to limit the length, but using the uint64 segments lets us interpret the java `long` as unsigned for bit manipulations and keep varint packing efficiency; though we do have to be somewhat careful not to perform regular arithmetic operations on the `long` primitives.